aptos-move/block-executor/src/scheduler.rs

```diff
@@ -143,55 +203,126 @@ impl Scheduler {
         Self {
             num_txns,
             execution_idx: AtomicUsize::new(0),
-            validation_idx: AtomicUsize::new(0),
-            decrease_cnt: AtomicUsize::new(0),
-            num_active_tasks: AtomicUsize::new(0),
+            validation_idx: AtomicU64::new(0),
+            commit_state: Mutex::new((0, 0)),
             done_marker: AtomicBool::new(false),
             txn_dependency: (0..num_txns)
                 .map(|_| CachePadded::new(Mutex::new(Vec::new())))
                 .collect(),
             txn_status: (0..num_txns)
-                .map(|_| CachePadded::new(Mutex::new(TransactionStatus::ReadyToExecute(0, None))))
+                .map(|_| {
+                    CachePadded::new((
+                        RwLock::new(ExecutionStatus::ReadyToExecute(0, None)),
+                        RwLock::new(ValidationStatus::new()),
+                    ))
+                })
                 .collect(),
         }
     }
 
+    /// If successful, returns Some(TxnIndex), the index of committed transaction.
+    /// The current implementation has one dedicated thread to try_commit.
+    pub fn try_commit(&self) -> Option<TxnIndex> {
+        let mut commit_state_mutex = self.commit_state.lock();
+        let commit_state = commit_state_mutex.deref_mut();
+        let (commit_idx, commit_wave) = (&mut commit_state.0, &mut commit_state.1);
+
+        if *commit_idx == self.num_txns {
+            // All txns have been committed, the parallel execution can finish.
+            self.done_marker.store(true, Ordering::SeqCst);
+            return None;
+        }
+
+        if let Some(validation_status) = self.txn_status[*commit_idx].1.try_read() {
+            // Acquired the validation status lock, now try the status lock.
+            if let Some(status) = self.txn_status[*commit_idx].0.try_upgradable_read() {
```

Comment:
Nit: similar to Zekun's comment, .0 and .1 are not very readable.

Reply:
Added comments.